src/DateTimeParser.jl

```diff
@@ -347,74 +350,73 @@ function _parsedate(datetimestring::AbstractString; fuzzy::Bool=false,
-                        res["timezone"] = get(value)
-                    end
-                end
+    processymd!(res, ymd, mstridx, yearfirst=yearfirst, dayfirst=dayfirst)
 
-                i += 1
-            elseif !haskey(res, "tzname") &&
-                    ismatch(r"^\w+$", tokens[i]) && !(lowercase(tokens[i]) in JUMP)
-                # Timezone name?
-                oldindex = i
-
-                res["tzname"] = tokens[i]
-                while i+2 <= len && tokens[i+1] == "/"
-                    res["tzname"] = string(res["tzname"], "/", tokens[i+2])
-                    i += 2
-                end
-                i += 1
-                # Check for something like GMT+3, or BRST+3
-                if i+1 <= len && tokens[i] in ("+", "-") &&
-                        isdigit(tokens[i+1]) && length(tokens[i+1]) in (1,2) &&
-                        (i+2 > len || tokens[i+2] != ":")
-                    res["tzname"] = string(res["tzname"], tokens[i], tokens[i+1])
-                    i += 2
-                end
+    if !haskey(res, "timezone") && !isnull(tzoffset)
+        res["timezone"] = FixedTimeZone("local", get(tzoffset))
+    end
 
-                value = trytimezone(res["tzname"], timezone_infos)
-                if !isnull(value)
-                    res["timezone"] = get(value)
-                elseif fuzzy == true
-                    delete!(res, "tzname")
-                    i = oldindex+1
-                else
-                    error("Faild to parse date")
-                end
-            elseif !(lowercase(tokens[i]) in JUMP) && !fuzzy
-                error("Failed to parse date")
-            else
-                i += 1
-            end
+    return res
+end
+
+function _tryparsetimezone!(res::Dict, tokens::Array{ASCIIString}, i::Int, timezone_infos::Dict{AbstractString,TimeZone})
+    len = length(tokens)
+    tzname = ""
+    oldindex = i
+
+    if i <= len && tokens[i] == "("
+        i += 1
+    end
+
+    if i <= len && ismatch(r"^\w+$", tokens[i])
+        tzname = tokens[i]
+        while i+2 <= len && ismatch(r"^\w+$", tokens[i]) &&
+                (tokens[i+1] in ("/", "-", "_") || ismatch(r"^\d+$", tokens[i+1]))
+            tzname = string(tzname, tokens[i+1], tokens[i+2])
+            i += 2
         end
+        i += 1
+    end
+
+    # Check for something like GMT+3, or BRST+3
+    if i+1 <= len && tokens[i] in ("+", "-") &&
+            isdigit(tokens[i+1]) && length(tokens[i+1]) in (1,2) &&
+            (i+2 > len || tokens[i+2] != ":")
+        tzname = string(tzname, tokens[i], tokens[i+1])
+        i += 2
+    end
+
+    if i <= len && tokens[i] == ")"
+        i += 1
     end
 
+    value = _tryparse(TimeZone, tzname, translation=timezone_infos)
+    if !isnull(value)
+        res["timezone"] = get(value)
+    else
+        i = oldindex
+    end
+
+    return i
+end
+
+function processymd!(res::Dict, ymd::Array, mstridx=-1; yearfirst=false, dayfirst=false)
```

Comment:
The reason why `processymd` does not just return a Year, Month and Day is because there isn't always a Year, Month and Day to return.